Kernel/FileSystem/FATFileSystem.cpp

```diff
@@ -0,0 +1,357 @@
+        return TRY(KString::try_create(filename.string_view()));
+    } else {
+        StringBuilder filename;
+        for (auto& lfn_entry : lfn_entries) {
+            filename.append(lfn_entry.characters1[0]);
+            filename.append(lfn_entry.characters1[1]);
+            filename.append(lfn_entry.characters1[2]);
+            filename.append(lfn_entry.characters1[3]);
+            filename.append(lfn_entry.characters1[4]);
+            filename.append(lfn_entry.characters2[0]);
+            filename.append(lfn_entry.characters2[1]);
+            filename.append(lfn_entry.characters2[2]);
+            filename.append(lfn_entry.characters2[3]);
+            filename.append(lfn_entry.characters2[4]);
+            filename.append(lfn_entry.characters2[5]);
+            filename.append(lfn_entry.characters3[0]);
+            filename.append(lfn_entry.characters3[1]);
+        }
+
+        return TRY(KString::try_create(byte_terminated_string(filename.string_view(), lfn_entry_text_termination)));
+    }
+
+    VERIFY_NOT_REACHED();
+}
+
+time_t FATInode::fat_date_time(u16 date, u16 time) const
+{
+    // Date format:
+    // - Year  7 bits
+    // - Month 4 bits
+    // - Day   5 bits
+
+    time_t year = (date >> 9) + first_fat_year_offset;
+    time_t month = (date >> 5) & 0b1111;
+    time_t day = date & 0b11111;
+
+    // Time format:
+    // - Hour 	5 bits
+    // - Minutes 6 bits
+    // - Seconds 5 bits (0 - 30, needs to be multiplied by two)
+
+    time_t hour = time >> 11;
+    time_t minute = (time >> 5) & 0b1111111;
+    time_t second = (time & 0b11111) * 2;
+
+    return second + minute * 60 + hour * 3600 + day + (24 * 3600) + month * (30 * 24 * 3600) + year * (365 * 24 * 3600);
+}
+
+StringView FATInode::byte_terminated_string(StringView string, u8 fill_byte)
+{
+    if (auto index = string.find_last_not_of(fill_byte); index.has_value())
+        return string.substring_view(0, index.value());
+    return string;
+}
+
+ErrorOr<size_t> FATInode::read_bytes_locked(off_t offset, size_t size, UserOrKernelBuffer& buffer, OpenFileDescription*) const
+{
+    dbgln_if(FAT_DEBUG, "FATFS: Reading inode {}: size: {} offset: {}", identifier().index(), size, offset);
+
+    // FIXME: Read only the needed blocks instead of the whole file
+    auto blocks = TRY(const_cast<FATInode&>(*this).read_block_list());
+    TRY(buffer.write(blocks->data() + offset, min(size, m_block_list.size() * fs().m_logical_block_size - offset)));
+
+    return min(size, m_block_list.size() * fs().m_logical_block_size - offset);
+}
+
+InodeMetadata FATInode::metadata() const
+{
+    return m_metadata;
+}
+
+ErrorOr<void> FATInode::traverse_as_directory(Function<ErrorOr<void>(FileSystem::DirectoryEntryView const&)> callback) const
+{
+    MutexLocker locker(m_inode_lock);
+
+    VERIFY(has_flag(m_entry.attributes, FATAttributes::Directory));
+
+    [[maybe_unused]] auto inode = TRY(const_cast<FATInode&>(*this).traverse([&callback](auto inode) -> ErrorOr<bool> {
```

Comment:
You could probably just drop the `[[maybe_unused]] auto inode = ` completely here.

Reply:
I can't, the compiler complains if I drop that.